src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
-        return null;
+    public T set(T value, int index) throws IndexOutOfBoundsException {
+        Node<T> nodeFound = findNodeByIndex(index);
+        T previousValue = nodeFound.item;
+        nodeFound.item = value;
+        return previousValue;
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public T remove(int index) throws IndexOutOfBoundsException {
+        Node<T> nodeToRemove = findNodeByIndex(index);
+        unlink(nodeToRemove);
+        size--;
+        return nodeToRemove.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.item == object
+                    || currentNode.item != null && currentNode.item.equals(object)) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void createFirstNode(Node<T> newNode) {
+        head = newNode;
+        tail = newNode;
+    }
+
+    private void linkBefore(Node<T> newNode, Node<T> nextNode) {
+        newNode.next = nextNode;
+        newNode.prev = nextNode.prev;
+        if (nextNode.prev != null) {
+            nextNode.prev.next = newNode;
+        } else {
+            head = newNode;
+        }
+        nextNode.prev = newNode;
+    }
+
+    private void linkAsLast(Node<T> newNode) {
+        newNode.prev = tail;
+        tail.next = newNode;
+        tail = newNode;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev == null && node.next == null) {
+            head = null;
+            tail = null;
+        } else if (node.prev == null) {
+            head = node.next;
+            head.prev = null;
+        } else if (node.next == null) {
+            tail = node.prev;
+            node.prev.next = null;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        validateIndex(index);
+        if (index + 1 <= size / 2) {
+            Node<T> currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            return currentNode;
+        } else {
+            Node<T> currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+            return currentNode;
+        }
+    }
+
+    private void validateIndex(int index) throws IndexOutOfBoundsException {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index
```

Comment:
is it RuntimeException? if yes, do we need to write `throws` in method signature?